src/main/java/org/opentripplanner/routing/trippattern/TripTimes.java

```diff
@@ -474,6 +478,66 @@ public boolean tripAcceptable(final State state0, final int stopIndex) {
         return true;
     }
 
+    /**
+     * Returns true if the current state chain contains a banned trip, or banned trip/stop index or banned trip sequence
+     */
+    public boolean tripOrTripSequenceIsBanned(State state0, final int stopIndex) {
+        final RoutingRequest options = state0.getOptions();
+
+        // check if an exact trip ID is banned
+        final BannedStopSet banned = options.bannedTrips.get(trip.getId());
+        if (banned != null && banned.contains(stopIndex)) {
+            return true;
+        }
+
+        // check if a trip ID sequence is banned, but first return false if there are not banned sequences
+        if (options.bannedTripSequences.size() == 0) {
+            return false;
+        }
+
+        // compile list of trips seen so far in states (use ArrayList constructor to ensure a mutable list is returned)
+        List<FeedScopedId> tripsInState = new ArrayList<>(Arrays.asList(trip.getId()));
+        State tripFindingState = state0;
+        Trip lastTrip = trip;
+        while (tripFindingState != null) {
+            if (tripFindingState.getBackEdge() != null) {
+                Trip trip = tripFindingState.getBackTrip();
+                if (trip != null && trip != lastTrip) {
+                    tripsInState.add(trip.getId());
+                    lastTrip = trip;
+                }
+            }
```

Comment:
There seems to be a lot of shared code here with: https://github.com/ibi-group/OpenTripPlanner/blob/08e034faace255e092211290220af3f60e553fa7/src/main/java/org/opentripplanner/routing/spt/GraphPath.java#L143-L149

Any potential for DRYing out?

Reply:
It is possible to consolidate code, but I think that abstraction might not be worth it for just this little block of code.